types/seat/wlr_seat.c

```diff
@@ -370,3 +370,36 @@ bool wlr_seat_validate_grab_serial(struct wlr_seat *seat, uint32_t serial) {
 	//	serial == seat->touch_state.grab_serial;
 	return true;
 }
+
+void wlr_serial_add(struct wlr_serial_ringset *set, uint32_t serial) {
+	if (set->count == 0 || set->data[set->end].max_incl + 1 != serial) {
+		set->count++;
+		if (set->count > WLR_SERIAL_RINGSET_SIZE) {
+			set->count = WLR_SERIAL_RINGSET_SIZE;
+		}
+		set->end = (set->end + 1) % WLR_SERIAL_RINGSET_SIZE;
+		set->data[set->end].min_incl = serial;
+		set->data[set->end].max_incl = serial;
+	} else {
+		set->data[set->end].max_incl = serial;
+	}
+}
+
+bool wlr_serial_maybe_valid(struct wlr_serial_ringset *set, uint32_t serial) {
+	for (int i = 0; i < set->count; i++) {
+		int j = (set->end - i + WLR_SERIAL_RINGSET_SIZE) % WLR_SERIAL_RINGSET_SIZE;
+		if (set->data[j].max_incl - serial > UINT32_MAX / 2) {
```

Comment:
I don't get these checks; what's the problem with simply checking `min_incl <= serial <= max_incl`? (assuming no serial overflows)

As it stands right now, you can have false negatives (consider `data[0].min_incl = 1, data[0].max_incl = MAX/2 + 2, serial = 1` on L392) and false positives with very new values (consider `data[0].max_incl = 1, data[0].min_incl = 1, serial = MAX/2 + 3` on L398), which doesn't match the purpose of the function. If you want to count all serials received before the start as valid, maybe add an explicit check?

Reply:
Serial overflows are still possible, so the code must be able to account for them. Also, as there were a few additional logic issues (sets without any entries admit all serials), I've replaced this function with something whose validation logic should not have any false positives.